lib/Base/System/ClientLists.vala

```diff
@@ -21,290 +21,169 @@
-   * Removes the access for an item from the KeyStorage.
-   */
-  internal override void remove_access (Session item) throws Error {
+    // Save server secrets
     try {
-      KeyStorage.remove_access (item.identifier);
+      PlatformEnum platform = PlatformEnum.for_server (server);
+      string token_label = @"Access Token for Server \"$(server.domain)\" on $(platform)";
+      KeyStorage.store_access (server.client_key, @"ck_$(server.identifier)", token_label);
+      string secret_label = @"Access Secret for Server \"$(server.domain)\" on $(platform)";
+      KeyStorage.store_access (server.client_secret, @"cs_$(server.identifier)", secret_label);
     } catch (Error e) {
       throw e;
     }
-  }
-
-}
-
-/**
- * Provides a list of all servers used by a Client.
- */
-public class Backend.ServerList : ClientList <Server> {
 
-  /**
-   * Creates a new instance of ServerList.
-   */
-  internal ServerList () {
-    Object ();
+    // Add the server to the list
+    add_item (server);
   }
 
   /**
-   * Adds the access for an item from the KeyStorage.
+   * Removes a server from the server list.
+   *
+   * @param server The server to be removed.
+   *
+   * @throws Error Errors when removing the secrets to the KeyStorage.
    */
-  internal override void add_access (Server item) throws Error {
+  internal void remove_server (Server server) throws Error {
+    // Ignore if server don't exists in list
+    if (! find (server, null)) {
+      return;
+    }
+
+    // Remove server secrets
     try {
-      PlatformEnum platform = PlatformEnum.for_server (item);
-      string token_label = @"Access Token for Server \"$(item.domain)\" on $(platform)";
-      KeyStorage.store_access (item.client_key, @"ck_$(item.identifier)", token_label);
-      string secret_label = @"Access Secret for Server \"$(item.domain)\" on $(platform)";
-      KeyStorage.store_access (item.client_secret, @"cs_$(item.identifier)", secret_label);
+      KeyStorage.remove_access (@"ck_$(server.identifier)");
+      KeyStorage.remove_access (@"cs_$(server.identifier)");
     } catch (Error e) {
       throw e;
     }
+
+    // Remove the server from the list
+    remove_item (server);
   }
 
   /**
-   * Removes the access for an item from the KeyStorage.
+   * Used to compares two iterators in the list when sorting.
+   *
+   * @param a The first iterator to compare.
+   * @param b The second iterator to compare.
+   *
+   * @return How the iterators are sorted (positive when a before b, negative when b before a).
    */
-  internal override void remove_access (Server item) throws Error {
-    try {
-      KeyStorage.remove_access (@"ck_$(item.identifier)");
-      KeyStorage.remove_access (@"cs_$(item.identifier)");
-    } catch (Error e) {
-      throw e;
-    }
+  protected override int sort_func (SequenceIter<Server> a, SequenceIter<Server> b) {
+    Server server_a = a.get ();
+    Server server_b = b.get ();
+    return strcmp (server_a.domain, server_b.domain);
```

Comment:
Are we assuming just one server per domain? Can't remember whether creation enforces that.

If not then we might want a secondary comparison for sort stability.

Reply:
`SessionAuth` uses `find_with_needle` to look for an existing server for the domain, and only creates a new server if none was found, so comparing by domain should be enough.